src/backend/optimizer/plan/setrefs.c

```diff
@@ -396,23 +412,31 @@ set_plan_references(PlannerInfo *root, Plan *plan)
 	 */
 	add_rtes_to_flat_rtable(root, false);
 
+	int	glen = list_length(glob->finalrtable);
 	/*
 	 * Adjust RT indexes of PlanRowMarks and add to final rowmarks list
 	 */
 	foreach(lc, root->rowMarks)
 	{
 		PlanRowMark *rc = lfirst_node(PlanRowMark, lc);
-		PlanRowMark *newrc;
 
-		/* flat copy is enough since all fields are scalars */
-		newrc = (PlanRowMark *) palloc(sizeof(PlanRowMark));
-		memcpy(newrc, rc, sizeof(PlanRowMark));
+		if (plan->type != T_ShareInputScan && !rowmark_list_has_rowmark(glob->finalrowmarks, rc)) 
```

Comment:
`&&` now excludes SELECT FOR UPDATE case. And any other case with shared scan.
Maybe just leave only rowmark_list_has_rowmark check? But firsly answer the quesion in previous review comment.